quote-book/src/synchronized_quote_book.rs

```diff
@@ -3,27 +3,100 @@
+use mc_ledger_db::{Error as LedgerError, Ledger};
 use mc_transaction_extra::SignedContingentInput;
-use std::ops::RangeBounds;
 
+use std::{
+    ops::RangeBounds,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+    thread::{Builder as ThreadBuilder, JoinHandle},
+    time::Duration,
+};
+
+use crate::Msg;
+use mc_common::logger::{log, Logger};
+use postage::{broadcast::Sender, prelude::Sink};
 /// A wrapper for a quote book implementation that syncs quotes with the ledger
-#[derive(Clone)]
-pub struct SynchronizedQuoteBook<Q: QuoteBook, L: Ledger + Clone + 'static> {
+pub struct SynchronizedQuoteBook<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> {
     /// Quotebook being synchronized to the ledger by the Synchronized Quotebook
     quote_book: Q,
 
+    /// Ledger
     ledger: L,
+
+    /// Shared state
+    shared_state: Arc<Mutex<DbPollSharedState>>,
+
+    /// Join handle used to wait for the thread to terminate.
+    join_handle: Option<JoinHandle<()>>,
+
+    /// Stop request trigger, used to signal the thread to stop.
+    stop_requested: Arc<AtomicBool>,
 }
 
 impl<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> SynchronizedQuoteBook<Q, L> {
     /// Create a new Synchronized Quotebook
-    pub fn new(quote_book: Q, ledger: L) -> Self {
-        Self { quote_book, ledger }
+    pub fn new(quote_book: Q, ledger: L, msg_bus_tx: Sender<Msg>, logger: Logger) -> Self {
+        let ledger_clone = ledger.clone();
+        let quote_book_clone = quote_book.clone();
+        let shared_state = Arc::new(Mutex::new(DbPollSharedState::default()));
+        let thread_state = shared_state.clone();
+        let stop_requested = Arc::new(AtomicBool::new(false));
+        let thread_stop_requested = stop_requested.clone();
+        let join_handle = Some(
+            ThreadBuilder::new()
+                .name("LedgerDbFetcher".to_owned())
+                .spawn(move || {
+                    DbFetcherThread::start(
+                        ledger_clone,
+                        quote_book_clone,
+                        msg_bus_tx,
+                        0,
+                        thread_state,
+                        thread_stop_requested,
+                        logger,
+                    )
+                })
+                .expect("Could not spawn thread"),
+        );
+        Self {
+            quote_book,
+            ledger,
+            shared_state,
+            join_handle,
+            stop_requested,
+        }
+    }
+    pub fn get_current_block_index(&self) -> u64 {
+        let shared_state = self.shared_state.lock().expect("mutex poisoned");
+        shared_state.highest_processed_block_index
+    }
+
+    /// Stop and join the db poll thread
+    pub fn stop(&mut self) -> Result<(), ()> {
+        if let Some(join_handle) = self.join_handle.take() {
+            self.stop_requested.store(true, Ordering::SeqCst);
+            join_handle.join().map_err(|_| ())?;
```

Comment:
in other places, I think what we usually did is `.join().expect("error joining thread");` or something like this

Reply:
maybe i'm wrong -- i guess what you did here is the same as what's here: https://github.com/mobilecoinfoundation/mobilecoin/blob/2a7b65c8558ff05571ce0c331e8224b9f3a0056a/fog/view/server/src/db_fetcher.rs#L122

I'm fine to just be consistent

Reply:
i feel like it would be nice to at least log this error, whatever it is, that should be easy to do if we pass in a logger, right?

Reply:
it looks to me that we are not consistent about how we handle errors when joining:

```
$ git grep "join_handle.join()"
fog/ledger/server/src/db_fetcher.rs:            join_handle.join().map_err(|_| ())?;
fog/overseer/server/src/worker.rs:            join_handle.join().map_err(|_| ())?;
fog/view/server/src/db_fetcher.rs:            join_handle.join().map_err(|_| ())?;
fog/view/server/src/server.rs:            join_handle.join().map_err(|_| ())?;
mobilecoind/src/sync.rs:                        join_handle.join().expect("Failed joining worker thread");
mobilecoind/src/sync.rs:            join_handle.join().expect("SyncThread join failed");
peers/src/threaded_broadcaster.rs:            let _ = join_handle.join();
sgx/report-cache/untrusted/src/lib.rs:            join_handle.join().map_err(|_| Error::ThreadJoin)?;
watcher/src/bin/main.rs:            join_handle.join().map_err(|_| Error::ThreadJoin)?;
```

IMO `join_handle.join().expect("...")` is the nicest because it ensures that we'll actually see this error

Reply:
Thank you for looking into this to help me understand the right way to do this.